### DISCUSSION TITLE
v0.11.0

### DISCUSSION THREAD

**Discussion by aleokdev** [mattyhall/rs-tiled#260] [Announcements]
A really big feature update that adds support for virtual file systems, templates, text objects and more!

If you are planning to upgrade to the new version, check out the full release notes on the [CHANGELOG](https://github.com/mapeditor/rs-tiled/blob/current/CHANGELOG.md).
## What's Changed
* VFS support by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/199
* Enforce ResourceReader errors to be Send + Sync by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/206
* Add mut loader properties by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/207
* Remove `Layer` suffix from `LayerType` variants by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/203
* Formalize 0.11 branch by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/208
* Fix 0.11 build by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/215
* Fix & add tests for template object loading by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/227
* Rename `Chunk` & `ChunkWrapper` by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/231
* Merge current into next by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/229
* Add text objects by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/230
* Merge branch 'current' into next by @bjorn in https://github.com/mapeditor/rs-tiled/pull/246
* Merge current into next by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/251
* Finish 0.11 TODOs by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/253
* Updated lifetime of LayerTile returned from TileLayer by @Anti-Alias in https://github.com/mapeditor/rs-tiled/pull/254
* Merge current into next by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/258
* Next -> Current by @aleokdev in https://github.com/mapeditor/rs-tiled/pull/259

**Full Changelog**: https://github.com/mapeditor/rs-tiled/compare/v0.10.3...v0.11.0### 

<hr /><em>This discussion was created from the release <a href='https://github.com/mapeditor/rs-tiled/releases/tag/v0.11.0'>v0.11.0</a>.</em>

**Comment by B-Reif** [comment]
Do you have any details on how to migrate? I've got a case where a crate on v0.10 is currently using `Loader::load_tmx_map_from` with a standard `BufReader`. Now this method is gone and the `Loader` wants something that implements the crate's own `ResourceReader` trait. 

I'm not sure how to implement this trait. I just have some bytes and don't have an obvious candidate for the associated `Resource` type, Furthermore, it looks like the crate is just forwarding the Reader object as `Read` to the xml crate anyway, so I don't understand why this `ResourceReader` trait exists.

**Reply by aleokdev** [reply]
Sorry for the wait! GitHub didn't notify me until now...

> Do you have any details on how to migrate? I've got a case where a crate on v0.10 is currently using Loader::load_tmx_map_from with a standard BufReader. Now this method is gone and the Loader wants something that implements the crate's own ResourceReader trait.

If you just have some bytes, you can use the `MemoryReader` example in `ResourceReader`, adding your data to the struct and then returning it on the `load_from` fn: https://docs.rs/tiled/0.11.0/tiled/trait.ResourceReader.html

If you want to load from a BufReader instead, you can implement `ResourceReader` on a structure that holds the `BufReader`, and then just set `Resource = BufReader` as well as `Error = std::convert::Infallible`. Something like so:
```rs
struct Reader<R: Read>(BufReader<R>);

impl<R: Read> tiled::ResourceReader for MemoryReader<R> {
    type Resource = Self;
    type Error = std::convert::Infallible;

    fn read_from(&mut self, _path: &std::path::Path) -> std::result::Result<Self::Resource, Self::Error> {
        Ok(self)
    }
}
```

More examples are definitely required though, and I'll see what I can do in the following weeks about this. Sorry!

> Furthermore, it looks like the crate is just forwarding the Reader object as Read to the xml crate anyway, so I don't understand why this ResourceReader trait exists.

Although a `Read` object can deal with loading single files, the `ResourceReader` trait exists so that the user can specify what to load given a path. In your usecase this is irrelevant as you are probably just loading a single object, but once you have more than one (e.g. external tilesets, templates, etc) then you'll need to deal with loading stuff from multiple paths, which is what the trait is for.

**Reply by B-Reif** [reply]
Got it, thank you very much for clarifying and thanks for your work on a useful piece of software ❤️
